README.md
# Irrigation meter

Connect your IoT beaglebone black to a cheap flow meter and get graphes for current flow rate and accumalated water flow.<br/>

Result would look like this:</br>
1. [bounces](https://github.com/nivw/irrigation_meter/blob/master/demo_graphs/water.flow.bounces.open.faucet.png) is the current flow rate, and as you can see, when opening a faucet, the air is pushed out of the pipe causing bouncy flow rate.</br>
2. [flow counter](https://github.com/nivw/irrigation_meter/blob/master/demo_graphs/water.flow.count.png) , is the accumalated bounces throgh the sensor.</br>

## Hardware:
1. [Beaglebone black](https://beagleboard.org/black)</br>
2. [Water Flow sensor](http://www.seeedstudio.com/wiki/G3/4_Water_Flow_sensor)</br>

### Connection:<br/>
- connect the beaglebone pin P9_15 to the yellow sensor pin. Also supply +5V to the red sensor and GND.
- You need to use a pull-up for P5_15, either via using the beaglebone devicetree or by a 1K resistor from this pin to 3.3v+

## Software:
- install the image for debian 8.4 from [rcn-ee](http://beagleboard.org/latest-images)

### Notes:
- This installs the graphite packages and uses gunicorn to serve the graphes.
- To reduce flash wear due to many writes, I use the Ram to store the graphitefiles, and copy them to flash every 15 min.
- I use systemd service to run a simple python script to epoll the GPIO sensor
- Graphite webapp is available at port 8888 of the beaglebone IP.
	    

debian/changelog
irrigation-meter (0.1-1) jessie; urgency=low

  * Initial release.

 -- Niv Gal Waizer <nivw2008@fastmail.fm>  Tue, 1 Jul 2016 16:10:00 +0300


irrigation_meter/__init__.py
#!/usr/bin/python
import time
from socket import socket
import os
import select

#taken from irrigation-bone/wfs_config.py
gpio_base = '/sys/class/gpio'
gpio = '48' #P9_15
WATERFLOW_TIMEOUT = 60 #no water flow for a minute

gpio_path = os.path.join( gpio_base, 'gpio' + gpio )
gpio_value = os.path.join(gpio_path, 'value')

def main():
  accumulated_value = 0
  sock = socket()
  sock.connect( ('127.0.0.1', 2003) )
  previous_timestamp = timestamp = int( time.time() )
  with open(gpio_value) as gpio_fd:
    po = select.epoll()
    po.register(gpio_fd, select.EPOLLIN | select.EPOLLET)  
    while True:
      events = po.poll( WATERFLOW_TIMEOUT )
      timestamp = int( time.time() )
      gpio_fd.seek(0)
      value = gpio_fd.read().strip()
      if previous_timestamp == timestamp:
        accumulated_value +=1
      else:
        accumulated_value = 0
      message = 'water.flow.count {} {}\n' .format(accumulated_value, timestamp)
      print message
      sock.sendall(message)
      previous_timestamp = timestamp
  sock.close()

if __name__ == "__main__":
      main()


usr/sbin/use_tmpsfs_for_carbon.sh
#!/bin/sh
# This script is ment to allow running carbon from tmpfs
# instead of the flash to reduce wear
# another process will priodcly copy the files for the flash
# every 30 min
set -x
GRAPHITE_DIR="/var/lib/graphite"
GRAPHITE_TMP_DIR="${GRAPHITE_DIR}.ephemeral"
GRAPHITE_STOR_DIR="${GRAPHITE_DIR}.permanent"
SIZE="150m"

systemctl is-active carbon-cache.service && systemctl stop carbon-cache.service
if [ ! -d "$GRAPHITE_TMP_DIR" ] ;then
    mkdir ${GRAPHITE_TMP_DIR}
    mount -t tmpfs -o size=${SIZE} tmpfs ${GRAPHITE_TMP_DIR}
fi
if [ ! -d "$GRAPHITE_STOR_DIR" -a -d "$GRAPHITE_DIR" ] ;then
  mv ${GRAPHITE_DIR} ${GRAPHITE_STOR_DIR}
fi

[ -d "$GRAPHITE_DIR" ] || ln -s ${GRAPHITE_TMP_DIR} ${GRAPHITE_DIR}
echo 0 > /proc/sys/vm/swappiness
rsync --archive  ${GRAPHITE_STOR_DIR}/ ${GRAPHITE_TMP_DIR} || exit 1
#systemctl start carbon-cache.service


opt/graphite/webapp/graphite/local_settings.py
## Graphite local_settings.py
# Edit this file to customize the default Graphite webapp settings
#
# Additional customizations to Django settings can be added to this file as well

#####################################
# General Configuration #
#####################################
# Set this to a long, random unique string to use as a secret key for this
# install. This key is used for salting of hashes used in auth tokens,
# CRSF middleware, cookie storage, etc. This should be set identically among
# instances if used behind a load balancer.
SECRET_KEY = 'unobius_password'

# In Django 1.5+ set this to the list of hosts your graphite instances is
# accessible as. See:
# https://docs.djangoproject.com/en/dev/ref/settings/#std:setting-ALLOWED_HOSTS
ALLOWED_HOSTS = [ '*' ]

# Set your local timezone (Django's default is America/Chicago)
# If your graphs appear to be offset by a couple hours then this probably
# needs to be explicitly set to your local timezone.
#TIME_ZONE = 'America/Los_Angeles'
TIME_ZONE = 'Asia/Jerusalem'

# Override this to provide documentation specific to your Graphite deployment
#DOCUMENTATION_URL = "http://graphite.readthedocs.org/"

# Logging
#LOG_RENDERING_PERFORMANCE = True
#LOG_CACHE_PERFORMANCE = True
LOG_METRIC_ACCESS = True

# Enable full debug page display on exceptions (Internal Server Error pages)
DEBUG = True

# If using RRD files and rrdcached, set to the address or socket of the daemon
#FLUSHRRDCACHED = 'unix:/var/run/rrdcached.sock'

# This lists the memcached servers that will be used by this webapp.
# If you have a cluster of webapps you should ensure all of them
# have the *exact* same value for this setting. That will maximize cache
# efficiency. Setting MEMCACHE_HOSTS to be empty will turn off use of
# memcached entirely.
#
# You should not use the loopback address (127.0.0.1) here if using clustering
# as every webapp in the cluster should use the exact same values to prevent
# unneeded cache misses. Set to [] to disable caching of images and fetched data
#MEMCACHE_HOSTS = ['10.10.10.10:11211', '10.10.10.11:11211', '10.10.10.12:11211']
#DEFAULT_CACHE_DURATION = 60 # Cache images and data for 1 minute


#####################################
# Filesystem Paths #
#####################################
# Change only GRAPHITE_ROOT if your install is merely shifted from /opt/graphite
# to somewhere else
#GRAPHITE_ROOT = '/opt/graphite'

# Most installs done outside of a separate tree such as /opt/graphite will only
# need to change these three settings. Note that the default settings for each
# of these is relative to GRAPHITE_ROOT
#CONF_DIR = '/opt/graphite/conf'
#STORAGE_DIR = '/opt/graphite/storage'
#CONTENT_DIR = '/opt/graphite/webapp/content'

# To further or fully customize the paths, modify the following. Note that the
# default settings for each of these are relative to CONF_DIR and STORAGE_DIR
#
## Webapp config files
#DASHBOARD_CONF = '/opt/graphite/conf/dashboard.conf'
#GRAPHTEMPLATES_CONF = '/opt/graphite/conf/graphTemplates.conf'

## Data directories
# NOTE: If any directory is unreadable in DATA_DIRS it will break metric browsing
#WHISPER_DIR = '/opt/graphite/storage/whisper'
#RRD_DIR = '/opt/graphite/storage/rrd'
#DATA_DIRS = [WHISPER_DIR, RRD_DIR] # Default: set from the above variables
#LOG_DIR = '/opt/graphite/storage/log/webapp'
#INDEX_FILE = '/opt/graphite/storage/index'  # Search index file


#####################################
# Email Configuration #
#####################################
# This is used for emailing rendered Graphs
# Default backend is SMTP
#EMAIL_BACKEND = 'django.core.mail.backends.smtp.EmailBackend'
#EMAIL_HOST = 'localhost'
#EMAIL_PORT = 25
#EMAIL_HOST_USER = ''
#EMAIL_HOST_PASSWORD = ''
#EMAIL_USE_TLS = False
# To drop emails on the floor, enable the Dummy backend:
#EMAIL_BACKEND = 'django.core.mail.backends.dummy.EmailBackend'


#####################################
# Authentication Configuration #
#####################################
## LDAP / ActiveDirectory authentication setup
#USE_LDAP_AUTH = True
#LDAP_SERVER = "ldap.mycompany.com"
#LDAP_PORT = 389
#	OR
#LDAP_URI = "ldaps://ldap.mycompany.com:636"
#LDAP_SEARCH_BASE = "OU=users,DC=mycompany,DC=com"
#LDAP_BASE_USER = "CN=some_readonly_account,DC=mycompany,DC=com"
#LDAP_BASE_PASS = "readonly_account_password"
#LDAP_USER_QUERY = "(username=%s)"  #For Active Directory use "(sAMAccountName=%s)"
#
# If you want to further customize the ldap connection options you should
# directly use ldap.set_option to set the ldap module's global options.
# For example:
#
#import ldap
#ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_ALLOW)
#ldap.set_option(ldap.OPT_X_TLS_CACERTDIR, "/etc/ssl/ca")
#ldap.set_option(ldap.OPT_X_TLS_CERTFILE, "/etc/ssl/mycert.pem")
#ldap.set_option(ldap.OPT_X_TLS_KEYFILE, "/etc/ssl/mykey.pem")
# See http://www.python-ldap.org/ for further details on these options.

## REMOTE_USER authentication. See: https://docs.djangoproject.com/en/dev/howto/auth-remote-user/
USE_REMOTE_USER_AUTHENTICATION = True

# Override the URL for the login link (e.g. for django_openid_auth)
#LOGIN_URL = '/account/login'


##########################
# Database Configuration #
##########################
# By default sqlite is used. If you cluster multiple webapps you will need
# to setup an external database (such as MySQL) and configure all of the webapp
# instances to use the same database. Note that this database is only used to store
# Django models such as saved graphs, dashboards, user preferences, etc.
# Metric data is not stored here.
#
# DO NOT FORGET TO RUN 'manage.py syncdb' AFTER SETTING UP A NEW DATABASE
#
# The following built-in database engines are available:
#  django.db.backends.postgresql          # Removed in Django 1.4
#  django.db.backends.postgresql_psycopg2
#  django.db.backends.mysql
#  django.db.backends.sqlite3
#  django.db.backends.oracle
#
# The default is 'django.db.backends.sqlite3' with file 'graphite.db'
# located in STORAGE_DIR
#
#DATABASES = {
#    'default': {
#        'NAME': '/opt/graphite/storage/graphite.db',
#        'ENGINE': 'django.db.backends.sqlite3',
#        'USER': '',
#        'PASSWORD': '',
#        'HOST': '',
#        'PORT': ''
#    }
#}
#


#########################
# Cluster Configuration #
#########################
# (To avoid excessive DNS lookups you want to stick to using IP addresses only in this entire section)
#
# This should list the IP address (and optionally port) of the webapp on each
# remote server in the cluster. These servers must each have local access to
# metric data. Note that the first server to return a match for a query will be
# used.
#CLUSTER_SERVERS = ["10.0.2.2:80", "10.0.2.3:80"]

## These are timeout values (in seconds) for requests to remote webapps
#REMOTE_STORE_FETCH_TIMEOUT = 6   # Timeout to fetch series data
#REMOTE_STORE_FIND_TIMEOUT = 2.5  # Timeout for metric find requests
#REMOTE_STORE_RETRY_DELAY = 60    # Time before retrying a failed remote webapp
#REMOTE_STORE_USE_POST = False    # Use POST instead of GET for remote requests
#REMOTE_FIND_CACHE_DURATION = 300 # Time to cache remote metric find results

## Prefetch cache
# set to True to fetch all metrics using a single http request per remote server
# instead of one http request per target, per remote server.
# Especially useful when generating graphs with more than 4-5 targets or if
# there's significant latency between this server and the backends. (>20ms)
#REMOTE_PREFETCH_DATA = False

# During a rebalance of a consistent hash cluster, after a partition event on a replication > 1 cluster,
# or in other cases we might receive multiple TimeSeries data for a metric key.  Merge them together rather
# that choosing the "most complete" one (pre-0.9.14 behaviour).
#REMOTE_STORE_MERGE_RESULTS = True

## Remote rendering settings
# Set to True to enable rendering of Graphs on a remote webapp
#REMOTE_RENDERING = True
# List of IP (and optionally port) of the webapp on each remote server that
# will be used for rendering. Note that each rendering host should have local
# access to metric data or should have CLUSTER_SERVERS configured
#RENDERING_HOSTS = []
#REMOTE_RENDER_CONNECT_TIMEOUT = 1.0

# If you are running multiple carbon-caches on this machine (typically behind a relay using
# consistent hashing), you'll need to list the ip address, cache query port, and instance name of each carbon-cache
# instance on the local machine (NOT every carbon-cache in the entire cluster). The default cache query port is 7002
# and a common scheme is to use 7102 for instance b, 7202 for instance c, etc.
#
# You *should* use 127.0.0.1 here in most cases
#CARBONLINK_HOSTS = ["127.0.0.1:7002:a", "127.0.0.1:7102:b", "127.0.0.1:7202:c"]
#CARBONLINK_TIMEOUT = 1.0
# Using 'query-bulk' queries for carbon
# It's more effective, but python-carbon 0.9.13 (or latest from 0.9.x branch) is required
# See https://github.com/graphite-project/carbon/pull/132 for details
#CARBONLINK_QUERY_BULK = False

#####################################
# Additional Django Settings #
#####################################
# Uncomment the following line for direct access to Django settings such as
# MIDDLEWARE_CLASSES or APPS
#from graphite.app_settings import *



irrigation_meter/log_water_sensor.py
#!/usr/bin/python
import time
from socket import socket
import os
import select

#taken from irrigation-bone/wfs_config.py
gpio_base = '/sys/class/gpio'
gpio = '48' #P9_15
WATERFLOW_TIMEOUT = 60 #no water flow for a minute

gpio_path = os.path.join( gpio_base, 'gpio' + gpio )
gpio_value = os.path.join(gpio_path, 'value')

def main():
  bounches_sum = bounches_during_sec = 0
  sock = socket()
  sock.connect( ('127.0.0.1', 2003) )
  previous_timestamp = timestamp = int( time.time() )
  with open(gpio_value) as gpio_fd:
    po = select.epoll()
    po.register(gpio_fd, select.EPOLLIN | select.EPOLLET)  
    while True:
      events = po.poll( WATERFLOW_TIMEOUT )
      timestamp = int( time.time() )
      gpio_fd.seek(0)
      value = gpio_fd.read().strip()
      if previous_timestamp == timestamp:
        bounches_during_sec +=1
        message = 'water.flow.bounces {} {}\n' .format(bounches_during_sec, timestamp)
        sock.sendall(message)
      else:
        bounches_sum += bounches_during_sec
        bounches_during_sec = 0
      message = 'water.flow.count {} {}\n' .format(bounches_sum, timestamp)
      #print message
      sock.sendall(message)
      previous_timestamp = timestamp
    po.unregister(gpio_fd)
  sock.close()

if __name__ == "__main__":
      main()
